File: backend/stripe_webhook.py
"""
Stripe Webhook Handler for SentientIQ
Handles subscription lifecycle events and updates user metadata
"""
import os
import json
import stripe
from fastapi import APIRouter, Request, HTTPException, Header
from typing import Optional
import logging

logger = logging.getLogger(__name__)

# Initialize Stripe
stripe.api_key = os.getenv("STRIPE_SECRET_KEY")
STRIPE_WEBHOOK_SECRET = os.getenv("STRIPE_WEBHOOK_SECRET")

# Price ID to plan mapping
PRICE_TO_PLAN = {
    os.getenv("STRIPE_PRO_PRICE_ID", "price_pro"): "pro",
    os.getenv("STRIPE_TEAM_PRICE_ID", "price_team"): "team",
    os.getenv("STRIPE_ENTERPRISE_PRICE_ID", "price_enterprise"): "enterprise",
}

webhook_router = APIRouter(prefix="/webhook", tags=["billing"])

def update_user_plan(customer_id: str, plan: str):
    """Update user's plan in your system (Clerk, DB, etc.)"""
    # TODO: Implement based on your auth system
    # For Clerk:
    # clerk_client.users.update_metadata(user_id, {"publicMetadata": {"plan": plan}})
    
    # For now, just log
    logger.info(f"Would update customer {customer_id} to plan {plan}")
    
    # If using a database, update here
    # db.execute("UPDATE users SET plan = ? WHERE stripe_customer_id = ?", (plan, customer_id))
    pass

@webhook_router.post("/stripe")
async def stripe_webhook(request: Request, stripe_signature: Optional[str] = Header(None)):
    """Handle Stripe webhook events"""
    
    if not STRIPE_WEBHOOK_SECRET:
        logger.warning("STRIPE_WEBHOOK_SECRET not configured")
        return {"ok": False, "error": "Webhook secret not configured"}
    
    # Get raw body
    try:
        payload = await request.body()
    except Exception as e:
        logger.error(f"Failed to read request body: {e}")
        raise HTTPException(status_code=400, detail="Invalid payload")
    
    # Verify signature
    try:
        event = stripe.Webhook.construct_event(
            payload, stripe_signature, STRIPE_WEBHOOK_SECRET
        )
    except ValueError:
        logger.error("Invalid payload")
        raise HTTPException(status_code=400, detail="Invalid payload")
    except stripe.error.SignatureVerificationError:
        logger.error("Invalid signature")
        raise HTTPException(status_code=400, detail="Invalid signature")
    
    # Handle the event
    event_type = event["type"]
    data = event["data"]["object"]
    
    logger.info(f"Processing Stripe event: {event_type}")
    
    if event_type == "checkout.session.completed":
        # New subscription created via Checkout
        session = data
        customer_id = session.get("customer")
        subscription_id = session.get("subscription")
        
        if subscription_id:
            # Get subscription details
            try:
                subscription = stripe.Subscription.retrieve(subscription_id)
                price_id = subscription["items"]["data"][0]["price"]["id"]
                plan = PRICE_TO_PLAN.get(price_id, "free")
                
                # Update user plan
                update_user_plan(customer_id, plan)
                
                logger.info(f"New subscription: {customer_id} -> {plan}")
            except Exception as e:
                logger.error(f"Failed to process new subscription: {e}")
    
    elif event_type == "customer.subscription.created":
        # Subscription created (backup handler)
        subscription = data
        customer_id = subscription.get("customer")
        price_id = subscription["items"]["data"][0]["price"]["id"]
        plan = PRICE_TO_PLAN.get(price_id, "free")
        
        update_user_plan(customer_id, plan)
        logger.info(f"Subscription created: {customer_id} -> {plan}")
    
    elif event_type == "customer.subscription.updated":
        # Plan changed
        subscription = data
        customer_id = subscription.get("customer")
        price_id = subscription["items"]["data"][0]["price"]["id"]
        plan = PRICE_TO_PLAN.get(price_id, "free")
        
        update_user_plan(customer_id, plan)
        logger.info(f"Subscription updated: {customer_id} -> {plan}")
    
    elif event_type in ["customer.subscription.deleted", "customer.subscription.canceled"]:
        # Subscription canceled/expired
        subscription = data
        customer_id = subscription.get("customer")
        
        update_user_plan(customer_id, "free")
        logger.info(f"Subscription canceled: {customer_id} -> free")
    
    elif event_type == "invoice.payment_failed":
        # Payment failed - optionally downgrade or notify
        invoice = data
        customer_id = invoice.get("customer")
        
        logger.warning(f"Payment failed for customer: {customer_id}")
        # Could send email notification or temporarily restrict access
    
    else:
        logger.info(f"Unhandled event type: {event_type}")
    
    return {"ok": True, "received": event_type}

# Optional: Customer portal management
@webhook_router.post("/create-portal-session")
async def create_portal_session(request: Request):
    """Create a Stripe Customer Portal session for subscription management"""
    
    body = await request.json()
    customer_id = body.get("customer_id")
    return_url = body.get("return_url", os.getenv("VITE_APP_URL", "http://localhost:5173") + "/billing")
    
    if not customer_id:
        raise HTTPException(status_code=400, detail="customer_id required")
    
    try:
        session = stripe.billing_portal.Session.create(
            customer=customer_id,
            return_url=return_url,
        )
        return {"url": session.url}
    except Exception as e:
        logger.error(f"Failed to create portal session: {e}")
        raise HTTPException(status_code=500, detail="Failed to create portal session")

# Health check for webhook
@webhook_router.get("/health")
def webhook_health():
    """Check if webhook handler is configured"""
    return {
        "ok": True,
        "configured": bool(STRIPE_WEBHOOK_SECRET),
        "plans_mapped": len(PRICE_TO_PLAN)
    }

File: backend/stripe-webhook.js
#!/usr/bin/env node
import express from 'express';
import Stripe from 'stripe';
import dotenv from 'dotenv';
import { createClient } from '@supabase/supabase-js';

dotenv.config({ path: '../.env' });

const app = express();
const stripe = new Stripe(process.env.STRIPE_SECRET_KEY);
const supabase = createClient(
  process.env.VITE_SUPABASE_URL,
  process.env.VITE_SUPABASE_ANON_KEY
);

const PORT = process.env.WEBHOOK_PORT || 8005;

// Map price IDs to plan names
const PRICE_TO_PLAN = {
  'price_1S4F69Ag3cgQ4qmbRsCSI8QI': 'pro',
  'price_1S4F6AAg3cgQ4qmbKvXh6D01': 'team',
  'price_1S4F6AAg3cgQ4qmbpmJWPg6J': 'enterprise'
};

// Raw body needed for webhook signature verification
app.use('/webhook', express.raw({ type: 'application/json' }));
app.use(express.json());

// Health check
app.get('/health', (req, res) => {
  res.json({ status: 'ok', service: 'stripe-webhook' });
});

// Stripe webhook endpoint
app.post('/webhook', async (req, res) => {
  const sig = req.headers['stripe-signature'];
  const webhookSecret = process.env.STRIPE_WEBHOOK_SECRET;

  let event;

  try {
    event = stripe.webhooks.constructEvent(req.body, sig, webhookSecret);
  } catch (err) {
    console.error('Webhook signature verification failed:', err.message);
    return res.status(400).send(`Webhook Error: ${err.message}`);
  }

  console.log(`📨 Received event: ${event.type}`);

  try {
    switch (event.type) {
      case 'checkout.session.completed': {
        const session = event.data.object;
        const customerId = session.customer;
        const subscriptionId = session.subscription;
        const clerkUserId = session.metadata?.clerk_user_id;

        if (!clerkUserId) {
          console.error('No clerk_user_id in session metadata');
          break;
        }

        // Get subscription details
        const subscription = await stripe.subscriptions.retrieve(subscriptionId);
        const priceId = subscription.items.data[0].price.id;
        const plan = PRICE_TO_PLAN[priceId] || 'free';

        // Update user subscription in Supabase
        const { error } = await supabase
          .from('user_subscriptions')
          .upsert({
            clerk_user_id: clerkUserId,
            stripe_customer_id: customerId,
            stripe_subscription_id: subscriptionId,
            plan,
            status: 'active',
            current_period_start: new Date(subscription.current_period_start * 1000).toISOString(),
            current_period_end: new Date(subscription.current_period_end * 1000).toISOString(),
            updated_at: new Date().toISOString()
          }, {
            onConflict: 'clerk_user_id'
          });

        if (error) {
          console.error('Error updating subscription:', error);
        } else {
          console.log(`✅ Subscription activated for user ${clerkUserId}: ${plan}`);
        }
        break;
      }

      case 'customer.subscription.updated': {
        const subscription = event.data.object;
        const priceId = subscription.items.data[0].price.id;
        const plan = PRICE_TO_PLAN[priceId] || 'free';

        // Update subscription in Supabase
        const { error } = await supabase
          .from('user_subscriptions')
          .update({
            plan,
            status: subscription.status,
            current_period_start: new Date(subscription.current_period_start * 1000).toISOString(),
            current_period_end: new Date(subscription.current_period_end * 1000).toISOString(),
            updated_at: new Date().toISOString()
          })
          .eq('stripe_subscription_id', subscription.id);

        if (error) {
          console.error('Error updating subscription:', error);
        } else {
          console.log(`✅ Subscription updated: ${subscription.id}`);
        }
        break;
      }

      case 'customer.subscription.deleted': {
        const subscription = event.data.object;

        // Mark subscription as canceled in Supabase
        const { error } = await supabase
          .from('user_subscriptions')
          .update({
            status: 'canceled',
            plan: 'free',
            updated_at: new Date().toISOString()
          })
          .eq('stripe_subscription_id', subscription.id);

        if (error) {
          console.error('Error canceling subscription:', error);
        } else {
          console.log(`❌ Subscription canceled: ${subscription.id}`);
        }
        break;
      }

      case 'invoice.payment_succeeded': {
        const invoice = event.data.object;
        console.log(`💰 Payment succeeded for invoice: ${invoice.id}`);
        
        // Log successful payment
        await supabase
          .from('payment_history')
          .insert({
            stripe_invoice_id: invoice.id,
            stripe_subscription_id: invoice.subscription,
            amount: invoice.amount_paid / 100, // Convert from cents
            currency: invoice.currency,
            status: 'succeeded',
            created_at: new Date().toISOString()
          });
        break;
      }

      case 'invoice.payment_failed': {
        const invoice = event.data.object;
        console.log(`⚠️ Payment failed for invoice: ${invoice.id}`);
        
        // Update subscription status
        await supabase
          .from('user_subscriptions')
          .update({
            status: 'past_due',
            updated_at: new Date().toISOString()
          })
          .eq('stripe_subscription_id', invoice.subscription);

        // Log failed payment
        await supabase
          .from('payment_history')
          .insert({
            stripe_invoice_id: invoice.id,
            stripe_subscription_id: invoice.subscription,
            amount: invoice.amount_due / 100,
            currency: invoice.currency,
            status: 'failed',
            created_at: new Date().toISOString()
          });
        break;
      }

      default:
        console.log(`Unhandled event type: ${event.type}`);
    }
  } catch (error) {
    console.error('Error processing webhook:', error);
    return res.status(500).send('Webhook processing error');
  }

  res.json({ received: true });
});

app.listen(PORT, () => {
  console.log(`🎯 Stripe webhook server running on port ${PORT}`);
  console.log(`📝 Webhook endpoint: http://localhost:${PORT}/webhook`);
  console.log(`\n⚡ Configure in Stripe Dashboard:`);
  console.log(`   1. Go to https://dashboard.stripe.com/webhooks`);
  console.log(`   2. Add endpoint: https://sentientiq.app/api/stripe/webhook`);
  console.log(`   3. Select events:`);
  console.log(`      - checkout.session.completed`);
  console.log(`      - customer.subscription.updated`);
  console.log(`      - customer.subscription.deleted`);
  console.log(`      - invoice.payment_succeeded`);
  console.log(`      - invoice.payment_failed`);
  console.log(`   4. Copy the signing secret to STRIPE_WEBHOOK_SECRET in .env`);
});